File: src/mcfoamy/scripts/controller_node.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String
from geometry_msgs.msg import Wrench, Pose, Twist
from std_srvs.srv import Trigger, TriggerResponse

from gazebo_example.controllers import PID, State
        
class ContollerNode(object):
    MAX_PUB_QUEUE = 10
  
    def __init__(self):
        self._start = [False]
        self._pid = None
        self._pose = Pose()
        self._twist = Twist()
        
    def init(self):
        self._wrench_pub = rospy.Publisher("external_wrench", 
                                           Wrench, 
                                           queue_size=self.MAX_PUB_QUEUE)
        self._pose_sub = rospy.Subscriber("pose", Pose, self._pose_callback)
        self._twist_sub = rospy.Subscriber("twist", Twist, self._twist_callback)
        self._start_service = rospy.Service('start_controller', Trigger, self._start_controller)
        self._pid = PID(5.0, 5.0, 0.0)
        
        return True
      
    def run(self):
        self._wait_for_trigger()
        
        frequency = 10
        rate = rospy.Rate(frequency) # 10hz
        
        while not rospy.is_shutdown():
            # ros::spinOnce not needed in Python
            
            self._wrench_pub.publish(self._compute_control_wrench(frequency))
            rate.sleep()
            
    def _wait_for_trigger(self):
      while not rospy.is_shutdown():
          if(self._start[0]):
              break
          rospy.loginfo_throttle(10, "Waiting for control start trigger")
          
    def _compute_control_wrench(self, frequency):
        desired_state = State(7.0, 0.0)
        measured_state = State(self._pose.position.z, self._twist.linear.z)
        force_z = self._pid.output(desired_state, measured_state, 1.0/frequency)
        
        command_wrench = Wrench()
        command_wrench.force.z = force_z
        
        return command_wrench
        
    def _pose_callback(self, data):
        self._pose = data
        
    def _twist_callback(self, data):
        self._twist = data
        
    def _start_controller(self, req):
        self._start[0] = True
        
        response = TriggerResponse()
        response.success = True
        return response

if __name__ == '__main__':
    try:
        rospy.init_node('controller_node_py', anonymous=True)
        node = ContollerNode()
        if not node.init():
            raise rospy.ROSInterruptException

        node.run()
        
    except rospy.ROSInterruptException:
        pass

File: src/mcfoamy/src/controller_node.cpp
#include "gazebo_example/controller_node.h"


namespace gazebo_example
{
  
const unsigned int ControllerNode::MAX_PUB_QUEUE = 10;
const unsigned int ControllerNode::MAX_SUB_QUEUE = 10;
  
  
ControllerNode::ControllerNode():
  start_(false)
  , node_handle("")
  , pid_(0.0, 0.0, 0.0)
{}

bool ControllerNode::init()
{
  wrench_pub_ = node_handle.advertise<geometry_msgs::Wrench>("external_wrench", MAX_PUB_QUEUE);
  pose_sub_ = node_handle.subscribe("pose", MAX_SUB_QUEUE, 
                                    &ControllerNode::poseCallback, this);
  twist_sub_ = node_handle.subscribe("twist", MAX_SUB_QUEUE, 
                                     &ControllerNode::twistCallback, this);
  start_service_ = node_handle.advertiseService("start_controller",
                                                &ControllerNode::start_controller,
                                                this);
  pid_ = controllers::PID(5.0, 5.0, 0.0);
  return true;
}

void ControllerNode::run()
{
  wait_for_trigger();
  
  const double frequency = 10.0; 
  ros::Rate loop_rate(frequency);
  
  while(ros::ok())
  {
    // Handle Callbacks
    ros::spinOnce();
    
    // Compute and publish controller output
    wrench_pub_.publish(compute_control_wrench(frequency));
    
    // Wait to maintain constant frequency
    loop_rate.sleep();
  }
}

void ControllerNode::wait_for_trigger()
{
  while(ros::ok())
  {
    ros::spinOnce();
    if(start_) {break;}
    ROS_INFO_THROTTLE(10, "Waiting for control start trigger");
  }
}

geometry_msgs::Wrench ControllerNode::compute_control_wrench(const double frequency)
{
  const controllers::State desired_state = {5.0, 0};
  const controllers::State measured_state = {pose_.position.z, twist_.linear.z};
  const double force_z = pid_.output(desired_state, measured_state, 1.0 / frequency);
  
  geometry_msgs::Wrench command_wrench;
  command_wrench.force.z = force_z;
  
  return command_wrench;
}


void ControllerNode::poseCallback(const geometry_msgs::Pose::ConstPtr& msg)
{
  pose_ = *msg;
}

void ControllerNode::twistCallback(const geometry_msgs::Twist::ConstPtr& msg)
{
  twist_ = *msg;
}

bool gazebo_example::ControllerNode::start_controller(std_srvs::Trigger::Request& req,
                                                      std_srvs::Trigger::Response& res)
{
  start_ = true;
  res.success = true;
}

} // gazebo_example namespace


int main(int argc, char **argv)
{
  ros::init(argc, argv, "controller_node");
  
  gazebo_example::ControllerNode node;
  if(!node.init())
  {
    return 0;
  }
  
  node.run();

  return 0;
}
